### ISSUE TITLE
Can't manually set subproject options that yield

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#5214]
The idea behind the `yield:` kwarg in `meson_options.txt` is to set the default value of the option, but apparently you cannot set the value *at all* on the command-line if it yields to an option in the super-project.

This was reported in gstreamer: https://gitlab.freedesktop.org/gstreamer/gst-build/issues/22#note_141592

We should always honour the value of a subproject option if the user explicitly sets it on the command-line, whether it yields or not.

**Comment by xclaesse** [comment]
As far as I know it has always been working like this, but I agree it's not the best behavior. I think yielding options should take as default value the value set on parent project, but still allow user to set a value explicitly on the subproject. This is how per-subproject built-in options works. https://mesonbuild.com/Builtin-options.html#specifying-options-per-subproject

**Comment by dabrain34** [comment]
yes it's a bit confusing and not intuiive. Is there any reason to keep it as it today ?

**Comment by xclaesse** [comment]
I think we should change it. I'm not sure what is the desired behavior regarding default_options. If a subproject does that:
```meson
project('some-subproject', default_options:['a_yielding_option=foo'])
```
and I configure with `-Da_yielding_option=bar` should it use the value set on parent project, or the default from subproject? In the per-subproject options case it would use the value from default_options, would be more consistent to do that as well, but I'm not sure that's the desired behavior.

**Comment by nirbheek** [comment]
I think the order for yielding options should be the same as non-yielding options: command-line args always override `subproject(..., default_options:)` which always override `project(..., default_options)`.

**Comment by tp-m** [comment]
I think command line options should always override the defaults from the `meson.build` or `meson_options.txt`.

And subproject/fallback `default_options` should override any option defaults that are "inherited" from command line (superproject option is set and subproject option yielded) or yielded `meson_options.txt` default - specific overrides general.

**Comment by xclaesse** [comment]
Ohh, did some quick test, seems I misunderstood my own per-subproject built-in option feature, lol :D

When a subproject does `project('foo', default_options: ['default_library=shared'])` and I configure with `-Ddefault_library=static` (setting on parent), the subproject is built as static. So command line option on parent override default on subproject. I think that's the right behavior and yielding project options should do the same.

**Comment by dabrain34** [comment]
yes but in the example -Dtests=enabled -Dsubproject:tests=disabled the test should be disabled as otherwise you have to enable the test one by one. I assume the subproject command line option should override everything.

**Comment by jpakkane** [comment]
The current behaviour is as it was originally designed and intentional. The reasoning behind it is the same as most things: do the simple thing and enforce it rather than try to support super maximal configuration of absolutely everything because it leads to incredibly complex monstrosities.

If yielding always yields then that is simple and understandable. If it sometimes yields and sometimes not based on a command line argument (that you may have added unintentionally) gets very confusing fast. Yes, this is at the expense of getting to specify a custom value even if the option is tagged as yielding.

Maybe overriding these could be done in the native file? At least then it would be configured in one unambiguous place rather than random command line arguments in the past.

**Comment by matthiasclasen** [comment]
Its a landmine. It means whenever I'm adding an option to my project, I must be aware off all the options of all the current and future subprojects, since one of them might be yielding to the name I choose.

**Comment by eli-schwartz** [comment]
I think a large part of it is the same reason options get frozen at first configure and then git pull changing the default is ignored even when you didn't specify a value at all and were defaulting on whatever the project picked.

Namely, it's difficult to track "intent" in large part because we don't know "where" any given option's value was originally set, so we don't know if that specific source (such as the project default_options) has changed. We would need to add additional tracking to the options dict, probably.

I'm not sure but I think the current pending option rework by @jpakkane (which he currently plans on merging right after the stable release of 1.7.0) may either solve this or make it easy to solve?

**Comment by jpakkane** [comment]
The option rewrite should not change how project options are handled, only builtin ones.

**Comment by bonzini** [comment]
> The option rewrite should not change how project options are handled, only builtin ones.

Surprise! :) This was fixed by 1.8, and is documented as of 1.8.3/1.9.
